pkg/controller/controller.go

```diff
@@ -191,25 +196,43 @@ func (c *Controller) onDelete(obj interface{}) {
 	}
 }
 
-func (c *Controller) onAddActionSet(as *crv1alpha1.ActionSet) error {
-	as, err := c.crClient.CrV1alpha1().ActionSets(as.GetNamespace()).Get(context.TODO(), as.GetName(), v1.GetOptions{})
+func (c *Controller) createBpCache(ctx context.Context, ns string) (map[string]*crv1alpha1.Blueprint, error) {
+	bps, err := c.crClient.CrV1alpha1().Blueprints(ns).List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	bpMap := make(map[string]*crv1alpha1.Blueprint)
+	for _, bp := range bps.Items {
+		bpMap[bp.Name] = bp
+	}
```

Comment:
I don't see where this is validated, but if I'm not mistaken according to the code actions in an actionset may have associations with different blueprints.
Using an informer will add need to wait after adding a blueprint and before adding an actions due to cache resync. Do we want such behavior?

Reply:
not really @ed-shilo. An actionset can point to just one blueprint.

Reply:
created blueprints not being available immediately *might* be a concern.

Reply:
The blueprint informer watches for new blueprints and updates its indexer accordingly. Try add a new blueprint, the blueprint informer should respond immediately. I am not sure why we are using the low level `cache` directly [here](https://github.com/kanisterio/kanister/blob/a83a548072c420d676c47e1daac833cbe46cfcb4/pkg/controller/controller.go#L105-L122), but it looks like we are already using client-go caching mechanism.

Re-sync applies to _existing_ resources, when client-go tries to reconcile any differences between the cache and the data in its store.

@ed-shilo You are right about the 1:n mapping between actionset and blueprint. It looks like an actionset can have [multiple actions](https://github.com/kanisterio/kanister/blob/a83a548072c420d676c47e1daac833cbe46cfcb4/pkg/apis/cr/v1alpha1/types.go#L77-L80), each of which can point to whichever blueprint it wants, even though downstream normally associate one actionset to one blueprint. Not sure why the API is designed that way.

Reply:
As far as I see, each action inside of actionset has separate blueprint field. Is it formally assumed that it should have the same value in case of multiple actions?

Reply:
@ihcsim tried to use lister for getting blueprints, but as I see a newly created blueprint goes into cache after some small time. In tests this is observed in failing of attempts to add an action sets linked with an added blueprint due to cache miss. Are we OK with this?

Reply:
You are right, even though we don't use different blueprints for separate actions in an actionset, it does look like the code allows it.

Reply:
> Is it formally assumed that it should have the same value in case of multiple actions?

No, we can't assume that anymore. If the API let user do multiple blueprints, we have to support it.

Re: missing cache, did you use the informers lister or the [cache lister](https://pkg.go.dev/k8s.io/client-go/tools/cache#Lister). Looks like our code is using the cache one.

Reply:
@ihcsim I used the lister from [the link](https://github.com/kanisterio/kanister/blob/c2a8e32e36a80a26d424f99699422d972667ec98/pkg/client/informers/externalversions/cr/v1alpha1/blueprint.go#L37-L40) you proposed previosly. As far as I see, it uses caching.